src/types/UiaTextRangeBase.cpp

```diff
@@ -372,24 +637,174 @@ try
+        return _getAttrVerificationFn(attributeId, *pRetVal);
+    }
+    case UIA_UnderlineStyleAttributeId:
+    {
+        pRetVal->vt = VT_I4;
+        if (attr.IsDoublyUnderlined())
+        {
+            pRetVal->lVal = TextDecorationLineStyle_Double;
+        }
+        else if (attr.IsUnderlined())
+        {
+            pRetVal->lVal = TextDecorationLineStyle_Single;
+        }
+        else
+        {
+            pRetVal->lVal = TextDecorationLineStyle_None;
+        }
+        return _getAttrVerificationFn(attributeId, *pRetVal);
+    }
+    default:
+        // This attribute is not supported.
+        pRetVal->vt = VT_UNKNOWN;
+        UiaGetReservedNotSupportedValue(&pRetVal->punkVal);
+        return nullptr;
+    }
+}
+
+IFACEMETHODIMP UiaTextRangeBase::GetAttributeValue(_In_ TEXTATTRIBUTEID attributeId,
                                                    _Out_ VARIANT* pRetVal) noexcept
+try
 {
     RETURN_HR_IF(E_INVALIDARG, pRetVal == nullptr);
+    VariantInit(pRetVal);
 
-    if (textAttributeId == UIA_IsReadOnlyAttributeId)
+    // AttributeIDs that require special handling
+    switch (attributeId)
+    {
+    case UIA_FontNameAttributeId:
+    {
+        pRetVal->vt = VT_BSTR;
+        pRetVal->bstrVal = SysAllocString(_pData->GetFontInfo().GetFaceName().data());
+        UiaTracing::TextRange::GetAttributeValue(*this, attributeId, *pRetVal);
+        return S_OK;
+    }
+    case UIA_IsReadOnlyAttributeId:
     {
         pRetVal->vt = VT_BOOL;
         pRetVal->boolVal = VARIANT_FALSE;
+        UiaTracing::TextRange::GetAttributeValue(*this, attributeId, *pRetVal);
+        return S_OK;
     }
-    else
+    default:
+        break;
+    }
+
+    // AttributeIDs that are exposed via TextAttribute
+    std::function<bool(const TextAttribute&)> checkIfAttrFound;
+    try
+    {
+        checkIfAttrFound = _getAttrVerificationFnForFirstAttr(attributeId, pRetVal);
+        if (!checkIfAttrFound)
+        {
+            // The AttributeID is not supported.
+            pRetVal->vt = VT_UNKNOWN;
+            UiaTracing::TextRange::GetAttributeValue(*this, attributeId, *pRetVal, UiaTracing::AttributeType::Unsupported);
+            return UiaGetReservedNotSupportedValue(&pRetVal->punkVal);
+        }
+    }
+    catch (...)
+    {
+        LOG_HR(wil::ResultFromCaughtException());
+        UiaTracing::TextRange::GetAttributeValue(*this, attributeId, *pRetVal, UiaTracing::AttributeType::Error);
+        return E_INVALIDARG;
+    }
+
+    if (IsDegenerate())
     {
+        // Unlike a normal text editor, which applies formatting at the caret,
+        // we don't know what attributes are written at a degenerate range.
+        // So let's return UiaGetReservedMixedAttributeValue.
+        // Source: https://docs.microsoft.com/en-us/windows/win32/api/uiautomationcore/nf-uiautomationcore-itextrangeprovider-getattributevalue

```

Comment:
I'm not sure I fully understand. You're saying that the degenerate range... which is selecting nothing except a single insertion point at the cursor... has no formatting yet. Yes? But I don't understand why `UiaGetReservedMixedAttributeValue` is the appropriate response and the linked doc didn't really clear that up for me either.

Also... don't we have some sort of default formatting that could be presumed to be what's coming out for the next character unless it's changed?

And also... if someone has left-arrowed into the existing edit line... wouldn't we have formatting? Or is the range no longer degenerate as it would be width of 1?

Reply:
> I'm not sure I fully understand. You're saying that the degenerate range... which is selecting nothing except a single insertion point at the cursor... has no formatting yet. Yes? But I don't understand why UiaGetReservedMixedAttributeValue is the appropriate response and the linked doc didn't really clear that up for me either.

This is one of those cases where terminal emulators and UIA don't play nice.

UIA's philosophy here is that you're on a text document (like MS Word). So there, you can actually apply formatting to the insertion point.

Terminal emulators don't work that way. The formatting is determined on each cell by the shell on the other side. So we don't have a context as to what the formatting is for a 0-cell point.

`GetAttributeValue` can return one of three things:
- an actual answer based on the attribute ID provided (i.e. IsItalic --> boolean, FontName --> string)
- `...MixedAttributeValue`: we support this, but we the value may vary
- `...NotSupportedValue`: we don't support this

I'm concerned that if I return `...NotSupported`, a screen reader may interpret that as "this attribute is not supported".

> Also... don't we have some sort of default formatting that could be presumed to be what's coming out for the next character unless it's changed?

I guess that would be `TextBuffer::GetCurrentAttributes`? If that's the case, I could move the degenerate check earlier and do something like this:
```cpp
if (IsDegenerate())
{
   const auto currentAttr{_pData->GetTextBuffer().GetCurrentAttributes()};
   checkIfAttrFound = _getAttrVerificationFnForAttr(attributeId, pRetVal, currentAttr);
   if (!checkIfAttrFound)
   {
      return UiaGetReservedNotSupportedValue();
   }
   UiaTracing::TextRange::GetAttributeValue(*this, attributeId, *pRetVal);
   return S_OK;
}
```

> And also... if someone has left-arrowed into the existing edit line... wouldn't we have formatting? Or is the range no longer degenerate as it would be width of 1?

What do you mean by "left-arrowed"? Like, in the input buffer? We'd have a degenerate range there. So I think you're right, using `GetCurrentAttributes` above might be the way to go.

Reply:
Yes, that is what I meant by "left-arrowed".

I didn't understand that `MixedAttributeValue` meant that "we support it but it may vary". I don't really know what a screen reader would do with that information. I'd imagine it would make more follow-up calls that might return whatever it wants or just move on (as opposed to giving up with `NotSupportedValue`). If it works, I'm fine with it.

But I do also think that `GetCurrentAttributes` might be a "better" answer as it's concrete and not just "yes, yes, but I cannot tell you now".